Protocol switching for cleartext HTTP

GallaFrancesco opened the pull request:
This PR means to introduce a first proposal for protocol switching in cleartext HTTP. The aim is to discuss the proposed changes so that by the end of the month they can be polished and hopefully merged. The HTTPS protocol switching uses ALPN, so most of it should be independent from this PR.

What has been done here:

1. Introduced a `switchProtocol` overload in `server.d` which is able to accept a `connection_handler` as a template parameter, send a 101 response and initialize the connection for the handler. For now, the initialization uses a dummy `HTTP2ConnectionStream` defined in `http2.d` which should be fully implemented as soon as compression and stream are introduced. The `startHTTP2Connection` method (below) is in charge of calling `switchProtocol` and passing the handler.

2. Added an "Upgrade" header check inside the HTTP/1 request handler, to detect an Upgrade request. The new code gets executed only if the protocol requested is "h2c" as per the RFC.

3. Introduced a `decode` method inside `HTTP2Settings` so that the payload received during the upgrade request can be used to set the proper settings. The method is consistent with the RFC in ignoring the upgrade request if it is not able to decode successfully the payload. Being fairly simple, the decoding process uses `std.Base64` and `std.bitmanip`, but I'd like to discuss better approaches that might suit the task better (and remain consistent with vibe.d in general).

4. The `startHTTP2Connection` is directly called from the HTTP1 handler and has the purpose of checking the SETTINGS payload validity, calling `res.switchProtocol` if the decoding is successful and otherwise closing the connection (even though connection management here is another point I would like to discuss).

5. A unittest was added along with a new commented RFC block. I left a lot of TODOs since they can be very helpful, especially with error handling, which hopefully I'll be able to complete by the end of the month.

With the proposed changes, the library is capable of receiving and processing correctly a HTTP/1 Upgrade request and reply with a SWITCHING_PROTOCOL response to the client. After that, both the client and the server should send each other a connection preface message, which in case of the server is a SETTINGS frame as defined in the RFC. For that reason, I decided to postpone its implementation until a Frame data structure is completed.

Note: the server protocol switching has been tested with `curl --http2 http://localhost:8080 -v` which has been very helpful since it displays every step of the client communication. The [http2 test suite](https://github.com/http2/http2-test) doesn't seem to be testing protocol switching but only the [main HTTP2 features](https://github.com/http2/http2-test/tree/master/lib/server/tests), so I assumed the best way was to have a simple client sending an upgrade request.

Edit: I had to comment some methods in HTTP2ConnectionStream since CI testing with LDC complained about them, I guess it makes no difference ATM as long as I have them as a memo.

s-ludwig commented on source/vibe/http/internal/http1.d:
```diff
@@ -319,6 +320,22 @@ private bool originalHandleRequest(InterfaceProxy!Stream http_stream, TCPConnect
+         * with value "h2c" (for cleartext)
+         * "h2" is ignored since it is used for TLS protocol switching (ALPN)
+         */
+        if("Upgrade" in req.headers && req.headers["Upgrade"] == "h2c" ) {
```
Avoids the double-lookup: `if (req.headers.get("Upgrade") == "h2c")`

s-ludwig commented on source/vibe/http/internal/http1.d:
```diff
@@ -319,6 +320,22 @@ private bool originalHandleRequest(InterfaceProxy!Stream http_stream, TCPConnect
+         * "h2" is ignored since it is used for TLS protocol switching (ALPN)
+         */
+        if("Upgrade" in req.headers && req.headers["Upgrade"] == "h2c" ) {
+            assert("HTTP2-Settings" in req.headers, "Upgrade request must include HTTP2-Settings");
```
Should throw an `HTTPStatusException` with `HTTPStatus.badRequest` instead of asserting.

For efficiency reasons, this should also do something like this:
```
auto psettings = "HTTP2-Settings" in req.headers;
enforceHTTP(psettings !is null, HTTPStatus.badRequest, "Upgrade request must include HTTP2-Settings");
auto h2settings = *psettings;
```

s-ludwig commented on source/vibe/http/internal/http1.d:
```diff
@@ -319,6 +320,22 @@ private bool originalHandleRequest(InterfaceProxy!Stream http_stream, TCPConnect
+
+            // try to start an HTTP/2 connection
+            // should return if settings is not correct
+            startHTTP2Connection(tcp_connection, h2settings, res);
```
Should probably `return` here and possibly even use `runTask` to decouple the HTTP/2 execution path from the HTTP/1 request loop.

s-ludwig commented on source/vibe/http/internal/http2.d:
```diff
@@ -112,42 +124,83 @@ HTTP2Setting http2Setting(HTTP2SettingID id, string name) {
         // must use labeled break w. static foreach
         assign: switch(id) {
-            default: assert(false, "Unsupported SETTINGS code.");
+            default: logWarn("Unsupported SETTINGS code:" ~ to!string(id));
```
Missing a `return;` here, otherwise falls through to the next case label.

s-ludwig commented on source/vibe/http/internal/http2.d:
```diff
@@ -112,42 +124,83 @@ HTTP2Setting http2Setting(HTTP2SettingID id, string name) {
+            uset = Decoder.decode(encodedSettings);
+            enforce!Base64Exception(uset.length % 6 == 0, "Invalid SETTINGS payload length");
+        } catch (Base64Exception) {
+            logWarn("Failed to decode SETTINGS payload");
```
Since this is triggered by the remove party, this should be lowered to `logDiagnostic`. I'd also always log the original exception message (just `e.msg`, `e.toString()` seems unnecessary here).

s-ludwig commented on source/vibe/http/internal/http2.d:
```diff
@@ -166,16 +219,211 @@ unittest {
+    // try decoding settings
+    if (settings.decode!Base64URL(h2settings))
+        switchRes.switchProtocol!(handleHTTP2Connection!HTTP2ConnectionStream)("h2c", settings);
+    else connection.close;
```
Should this reply with a 400 error before closing the connection?

s-ludwig commented on source/vibe/http/server.d:
```diff
@@ -2257,6 +2266,33 @@ struct HTTPServerResponseData {
+            {
+
+                // send SWITCHING_PROTOCOL request
+                statusCode = HTTPStatus.SwitchingProtocols;
```
Should be `switchingProtocols`, the upper-case version is deprecated.

GallaFrancesco commented:
Hi, thank you for the review! I corrected the code according to your suggestions, leaving out the `runTask` call of `startHTTP2Connection` since I figured it would be easier to add it later (as soon as streams are implemented). I also added a small method to quickly send a BadRequest error response in case the `HTTP2Settings.decode` function fails.

s-ludwig commented:
Okay, looks good now. BTW, how comfortable are you with interactive rebasing in general? Usually I'd aim for having each commit corresponding to a logical change, without any small fixup merge commits left, but it wouldn't make sense to enforce this if that would mean that you'd have to start fighting with git.

s-ludwig commented:
Can you squash "upgrade header field check" and "fix upgrade header field check" together? Otherwise looks good to merge!

GallaFrancesco commented:
Done! I must have missed it before. Thanks!

GallaFrancesco commented:
Thank you for your help! Tomorrow I'll commit the remaining changes regarding ALPN so that we can review them.
